### DISCUSSION TITLE
Why is this for semantics rule of `number = digit+ ("." digit+)?`?

### DISCUSSION THREAD

**Discussion by panlina** [ohmjs/ohm#347] [Help] [answered]
I have grammar:
```
	number = digit+ ("." digit+)?
```
and semantic rule:
```
	number: (integer, dot, decimal) => +(integer.sourceString + dot.sourceString + decimal.sourceString)
```

I have 2 questions:

* Why the arity should be 3, not 2? It is as if expressions in `?` are expanded in place.
* I expect `dot` and `decimal` referring to the dot and the decimal part respectively, but as I inspected, both `dot.sourceString` and `decimal.sourceString` are `".456"` with test case `123.456`, while `dot.children[0].sourceString` and `decimal.children[0].sourceString` are `"."` and `"456"` respectively. This looks strange to me, because there's no such thing whose only child is `.`/`456` while its source is `.456`. Why is it?

**Comment by pdubroy** [comment, marked as answer]
I'm really sorry that it took so long to answer this 😬

> It is as if expressions in `?` are expanded in place.

Yes, that's how Ohm works. The main reason for this is that it's unclear how you would write a separate semantic action for the `?` expression separately.

> * I expect `dot` and `decimal` referring to the dot and the decimal part respectively, but as I inspected, both `dot.sourceString` and `decimal.sourceString` are `".456"` with test case `123.456`, while `dot.children[0].sourceString` and `decimal.children[0].sourceString` are `"."` and `"456"` respectively. This looks strange to me, because there's no such thing whose only child is `.`/`456` while its source is `.456`. Why is it?

You're right that it doesn't really make sense. Unfortunately, this is something that we've never been able to find a good solution for. All nodes have a `sourceString` property, but I think it's not really meaningful for iteration nodes. One thing I considered is just throwing an exception if it's ever called.
